### ISSUE TITLE
Potential performance decrease with bufio on write

### ISSUE THREAD

**Issue by thedevop** [mochi-mqtt/server#322]
Up until now, we really do not use cl.Net.bconn for writes. A new buffer is created for writes every time and writes directly to cl.Net.Conn. This works well for large number of connections (as write buffer can be kept to a very small value) or large packet (as a new net.Buffers is created).

However with #321, although it can improve write when the buffer size is larger than packet size, it does come at a cost of larger memory footprint for large number of connections.

@x20080406 , @mochi-co thoughts?

**Comment by mochi-co** [comment]
@thedevop @x20080406 It's outside of the scale of my own use so I can't really provide any meaningful consideration either way. However, I wasn't aware of the memory impact of #321, so it sounds like we should address it. At minimum it might be worth putting this behind a configuration option.

**Comment by thedevop** [comment]
The memory impact is indirect. As now we'll be using the bufio of cl.Net.bconn for write, setting this value very low (ClientNetWriteBufferSize) will have impact on write performance where perviously didn't as it was not used. Increase it will add fixed memory footprint per connection.

**Comment by x20080406** [comment]
The current scenario I'm facing involves a very high QPS (Queries Per Second) for a single connection. 120 producers report 3000 records per second to  broker, and there have  120 clients slice as 4 groups to  subscribe all of the data. If we write data directly to the net.Conn, the network will becomes extremely significant. there has lots of dropped messages. but I write data to bufio. there does not have dropped message. So I think copy the data to memory incurs much lower overhead compared to the frequent network overhead.  If there have lots of clients connect to broker. It will need too much memory. But we can add an option in  mqtt.Options to control write data to bufio. @thedevop @mochi-co , how do you think about?

**Comment by thedevop** [comment]
@x20080406 , with the new change, the improvement occurs (as it holds off flushing) when:
1. there are more than one writes to a connection
2. the packet sizes are smaller than ClientNetWriteBufferSize

So I was thinking if we can combine the best of both world, where:
1. the write buffer is not statically allocated per connection
2. a buffer is dynamically allocated to combine small writes
3. the buffer is only held as long as there are pending writes

The follow is NOT final code, but a general direction:
```go
	n, err := func() (int64, error) {
		out := buf.Bytes()
		cl.Lock()
		defer cl.Unlock()
		cl.Net.outBuffer = append(cl.Net.outBuffer, out) // outBuffer net.Buffers

		// immediate flush if no packets in channel and write buffer is not empty
		if len(cl.State.outbound) == 0 && cl.Net.outBuffer != nil {
			_, err := cl.Net.outBuffer.WriteTo(cl.Net.Conn)
			if err != nil {
				return 0, err
			}
			cl.Net.outBuffer = nil
		}
		return int64(len(out)), nil
	}()
```
Improvements needed:
1. There maybe other buffer type that is more appropriate than net.Buffers
2. Error handling is similar to the new code, but the actual bytes written to wire is misleading
3. cl.State.outbound is not accurate when executed from multiple Go routine (but may not be an issue)

**Comment by x20080406** [comment]
> a buffer is dynamically allocated to combine small writes

This is a good idea. but the bconn has been allocated a unuesed buffer.  Is it should be remove, include the ClientNetWriteBufferSize?

In fact. in my prject the producer will not allocate the write buffer,  the subscriber will allocate it. I reset them it in auth.AuthOk. It's not a generic direction. :)

**Comment by thedevop** [comment]
The static write buff on bconn and potentially be removed. There should still be ClientNetWriteBufferSize or something similar, there should be a threshold to force a flush for those constantly busy client.

**Comment by thedevop** [comment]
@mochi-co, how would you like to consider packet written to buffer, but not flushed to wire? This impacts both the the metrics and hook OnPacketSent.

**Comment by werbenhu** [comment]
@x20080406 @mochi-co @thedevop I tested with the following client code, and it seems that the changes in #321, in my environment, caused the clients to miss a very large portion of messages.
```go
package main

import (
	"fmt"
	"log/slog"
	"os"
	"os/signal"
	"strconv"
	"sync/atomic"
	"syscall"

	paho "github.com/eclipse/paho.mqtt.golang"
)

func main() {

	sigs := make(chan os.Signal, 1)
	done := make(chan bool, 1)
	signal.Notify(sigs, os.Interrupt, syscall.SIGTERM)

	go func() {
		<-sigs
		done <- true
	}()

	var messageCount atomic.Int32
	opts := paho.NewClientOptions().
		AddBroker("tcp://127.0.0.1:1883").
		SetClientID("werbenhu-client-id")

	opts.OnConnect = func(client paho.Client) {
		slog.Info("Connected")
	}

	client := paho.NewClient(opts)
	if token := client.Connect(); token.Wait() && token.Error() != nil {
		panic(token.Error())
	}

	client.Subscribe("werbenhu/topic", 0, func(c paho.Client, m paho.Message) {
		messageCount.Add(1)
		fmt.Printf("messages: %d\n", messageCount.Load())
	})

	for i := 0; i < 5; i++ {
		go func(worker int) {
			jobs := 10000
			for j := 0; j < jobs; j++ {
				token := client.Publish("werbenhu/topic", 0, false, strconv.Itoa(worker*jobs+j))
				token.Wait()
			}
		}(i)
	}

	<-done
	client.Disconnect(100)
	fmt.Printf("exist\n")
}
```

**Comment by werbenhu** [comment]
```go
// immediate flush if no packets in channel and write buffer is not empty
if len(cl.State.outbound) == 0 && cl.Net.outBuffer != nil {
```
If cl.State.outbound is not empty here and the buffer is full, does it mean that the data is discarded?

**Comment by x20080406** [comment]
if there does not have 

> ```go
> // immediate flush if no packets in channel and write buffer is not empty
> if len(cl.State.outbound) == 0 && cl.Net.bconn.Writer.Buffered() > 0 {
> ```
> 
> If cl.State.outbound is not empty here and the buffer is full, does it mean that the data is discarded?

It means current packets is the end of outbound , the eventloop should flush data in the buffer to subscriber. data in buffer does not been discarded. 

if there has packet in outbound, eventloop  normally execute. and if data is full in the buffer, the buffer will flush data to subscriber.

broker cannot tolerate data being stored in the cache for a long time.  current goroutine has the lock `cl.Lock()`, so  this function is thread safe. flush operation is safe too.

**Comment by mochi-co** [comment]
In light of the findings by @thedevop and @werbenhu I have reverted the changes made in #321 until we have a consensus on the way forward. I like the direction @x20080406 is taking to increase performance, but we need to ensure it doesn't have unexpected side effects.

**Comment by werbenhu** [comment]
@deadprogram In theory, it does indeed seem like there are no problems. 

I'm running on a Linux server, and everything appears to be fine. However, when I run the broker on my Windows machine, this issue occurs. 

If I add this line of log print code to your code, more messages are missed, but with the previous code, there's no issue. 

It's a bit strange, and I haven't figured out where the problem is yet.

```go
// nb := net.Buffers{buf.Bytes()}
	n, err := func() (int64, error) {
		cl.Lock()
		defer cl.Unlock()

		// return nb.WriteTo(cl.Net.Conn)
		length, err := cl.Net.bconn.Write(buf.Bytes())
		if nil != err {
			return 0, err
		}

		fmt.Printf("out size:%d, buffered:%d\n", len(cl.State.outbound), cl.Net.bconn.Writer.Buffered())
		// immediate flush if no packets in channel and write buffer is not empty
		if len(cl.State.outbound) == 0 && cl.Net.bconn.Writer.Buffered() > 0 {
			if err = cl.Net.bconn.Flush(); nil != err {
				return 0, err
			}
		}

		return int64(length), nil
	}()
```

**Comment by x20080406** [comment]
1. there will be loss data in your testing environment is bound to occur because production exceeds consumption. It's unrelated to whether buffering is enabled.
2. you shouldn't print message in worker gorouine. the worker groutine will flush data to you screen monitor. the more prominent the issue becomes with larger payload,

```
client.Subscribe("werbenhu/topic", 0, func(c paho.Client, m paho.Message) {
		messageCount.Add(1)
	})
	go func() {
		for {
			fmt.Printf("messages: %d\n", messageCount.Load())
			time.Sleep(1 * time.Second)
		}
	}()
```

**Comment by werbenhu** [comment]
@x20080406 From the broker's stats, it appears that this is unrelated to the client. It should be a performance decrease in Client.WritePacket() caused by buffer usage, leading to an overflow of outbound messages that couldn't be processed in time. Consequently, when the outbound queue is full, the broker forcibly drops some of the messages.
```go
func (s *Server) publishToClient(cl *Client, sub packets.Subscription, pk packets.Packet) (packets.Packet, error) {
	if sub.NoLocal && pk.Origin == cl.ID {
		return pk, nil // [MQTT-3.8.3-3]
	}

	//....

	select {
	case cl.State.outbound <- &out:
		atomic.AddInt32(&cl.State.outboundQty, 1)
	default:
		atomic.AddInt64(&s.Info.MessagesDropped, 1)
		cl.ops.hooks.OnPublishDropped(cl, pk)
		cl.State.Inflight.Delete(out.PacketID) // packet was dropped due to irregular circumstances, so rollback inflight.
		cl.State.Inflight.IncreaseSendQuota()
		return out, packets.ErrPendingClientWritesExceeded
	}

	return out, nil
}
```

**Comment by werbenhu** [comment]
This is why the issue is occurring on my Windows computer because my Windows computer has poor performance.

**Comment by werbenhu** [comment]
@x20080406 @mochi-co @thedevop Using a bufio can increase throughput, but it also leads to a significant increase in memory and CPU usage. Currently, my use case doesn't involve this requirement, so I don't have a strong opinion on it. Perhaps we can add an option in mqtt.Options to control writing data to bufio.

**Comment by werbenhu** [comment]
> The current scenario I'm facing involves a very high QPS (Queries Per Second) for a single connection. 120 producers report 3000 *120 records per second to broker, and there have 120 clients slice as 4 groups to subscribe all of the data. If we write data directly to the net.Conn, the network will becomes extremely significant. there has lots of dropped messages. but I write data to bufio. there does not have dropped message. So I think copy the data to memory incurs much lower overhead compared to the frequent network overhead. If there have lots of clients connect to broker. It will need too much memory. But we can add an option in mqtt.Options to control write data to bufio. @thedevop @mochi-co , how do you think about?

@x20080406 I have doubts about whether there's a real improvement in QPS here. You should investigate the reason for the increase in QPS, whether it's due to the use of bufio or if most of the messages you've published are being dropped.

**Comment by thedevop** [comment]
@x20080406 , @werbenhu and @mochi-co , please take a look at PR #324, this tries to combine the best of both world. It will buffer the write when there are more writes in the queue or if the size of buffered msgs is < ClientNetWriteBufferSize. It uses bytes.Buffer that was already allocated.

|msg in queue | buffer exist | packet > max | action |
|--------------|------------|---------------|-------|
| false | false | n/a | write packet to conn
| false | true | n/a | write packet to buffer, write buffer to conn
| true | false | false | set buffer
| true | false | true | write packet to conn
| true | true | false | write packet to buffer
| true | true | true | write packet to buffer, write buffer to conn

So for not busy clients, it behaves almost like now, so we can avoid putting this behind a flag. Let me if we're okay with this direction.

@mochi-co, currently this only return success bytes written to conn, so the metrics for BytesSent should work without modification. However, this creates complication for PacketsSent and for OnPacketSent hook. Anything thoughts on if we want to treat write to buffer as success or only after successfully written to conn?

**Comment by thedevop** [comment]
Resolved with #324.
